source-firestore/main.go

```diff
@@ -0,0 +1,411 @@
+
+		var eg = new(errgroup.Group)
+		var schema schema_inference.Schema
+
+		eg.Go(func() error {
+			schema, err = schema_inference.Run(ctx, logEntry, docsCh)
+			if err != nil {
+				return fmt.Errorf("schema inference: %w", err)
+			}
+			return nil
+		})
+
+		for {
+			var doc, err = docs.Next()
+			if err == iterator.Done {
+				break
+			} else if err != nil {
+				return err
+			}
+			var data = doc.Data()
+			data[firestore.DocumentID] = doc.Ref.ID
+			docJson, err := json.Marshal(data)
+			docsCh <- docJson
+		}
+		close(docsCh)
+
+		if err := eg.Wait(); err != nil {
+			return err
+		}
+
+		catalog.Streams = append(catalog.Streams, airbyte.Stream{
+			Name:                    collection.ID,
+			JSONSchema:              schema,
+			SupportedSyncModes:      []airbyte.SyncMode{airbyte.SyncModeIncremental},
+			SourceDefinedCursor:     true,
+			SourceDefinedPrimaryKey: [][]string{{firestore.DocumentID}},
+		})
+	}
+
+	var encoder = airbyte.NewStdoutEncoder()
+	return encoder.Encode(airbyte.Message{
+		Type:    airbyte.MessageTypeCatalog,
+		Catalog: catalog,
+	})
+}
+
+// struct sent over channel to encoder
+type encDocument struct {
+	doc        []byte
+	streamName string
+	state      []byte
+}
+
+func doRead(args airbyte.ReadCmd) error {
+	var cfg config
+	var state State
+	var catalog airbyte.ConfiguredCatalog
+
+	if err := args.ConfigFile.Parse(&cfg); err != nil {
+		return err
+	} else if err := args.CatalogFile.Parse(&catalog); err != nil {
+		return err
+	} else if args.StateFile != "" {
+		if err := args.StateFile.Parse(&state); err != nil {
+			return err
+		}
+	}
+
+	var lastScan = time.UnixMilli(0)
+	if state.LastScan != "" {
+		var err error
+		lastScan, err = time.Parse(time.RFC3339, state.LastScan)
+		if err != nil {
+			return fmt.Errorf("parsing last scan checkpoint failed: %w", err)
+		}
+	}
+
+	var scanInterval, _ = time.ParseDuration("12h")
+	if cfg.ScanInterval != "" {
+		var err error
+		scanInterval, err = time.ParseDuration(cfg.ScanInterval)
+
+		if err != nil {
+			return fmt.Errorf("parsing scan interval failed: %w", err)
+		}
+	}
+
+	log.WithFields(log.Fields{
+		"last_scan":     lastScan,
+		"scan_interval": scanInterval,
+	}).Info("state check for scan")
+
+	ctx := context.Background()
+	sa := option.WithCredentialsJSON([]byte(cfg.CredentialsJSON))
+	app, err := firebase.NewApp(ctx, nil, sa)
+	if err != nil {
+		return err
+	}
+
+	client, err := app.Firestore(ctx)
+	if err != nil {
+		return err
+	}
+	defer client.Close()
+
+	// The channel where documents will be sent to from different streams
+	// this is here to ensure we write documents sequentially and avoid scrambled outputs
+	var docsCh = make(chan encDocument)
+
+	var eg = new(errgroup.Group)
+
+	for _, stream := range catalog.Streams {
+		var streamName = stream.Stream.Name
+
+		var collection = client.Collection(streamName)
+		eg.Go(func() error {
+			return fullScan(ctx, lastScan, scanInterval, collection, docsCh, catalog.Tail)
```

Comment:
If I'm understanding the code here correctly there's no synchronization between `listenForChanges` and `fullScan`, so it's possible in some cases [1] that `fullScan` might output an older version of some document _after_ a newer version was already output by `listenForChanges`. Does that sound correct? Do documents have some sort of timestamp/clock value such that this can be reconciled later on?

[1] For a concrete example, this seems like it could happen if the query results in `fullScan` represent a point-in-time snapshot of the dataset, or equivalently if there's enough buffering/latency between Firestore and the connector.

Reply:
Or is the idea that any particular document is unlikely to be constantly under contention, so at worst that document will be slightly stale until the next `fullScan` run some hours later?

Reply:
That is a valid concern, I was not very worried about it since I didn't expect there to be a high rate of edits, but that assumption doesn't always hold.

I'm also not sure whether the query for documents is a point-in-time snapshot or not, if it wasn't the current approach would work, but I couldn't find any information on that from Firestore docs.

It seems tricky to me to ensure that the full scan does not output any documents that have already been outputted by the listener, mostly because it can consume a lot of memory to hold all the documents that are being updated if there are a lot of them, although this memory can be cleaned up once the full scan is over, but this means on larger datasets, we are likely to run into memory issues: the full scan takes longer, there are potentially more updates, and so we have a faster growing list of updated documents which we need to hold onto until the full scan is over. 🤔